### PR TITLE
add n-back task

### PR DESCRIPTION
**Pull request by Floozutter**
The `jspsych-n-back.js` plugin is still incomplete, but I feel bad just leaving all this code lying around.

### CODE REVIEW

**Review by noah10 — CHANGES_REQUESTED** [review]

File: baseline/client/n-back/n-back.js [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+};
+
+NBack.cue = {
+    type: "html-keyboard-response",
+    stimulus: `<strong>+</strong>`,
+    choices: jsPsych.NO_KEYS,
+    trial_duration: 1000,
+};
+
+NBack.trial = (n, sequence, cue) => ({
```
[the remark this answers is not in the document]
Comment (Floozutter) [reply baseline/client/n-back/n-back.js]: Good catch! 47361ce02adabc458d0e6f4c329e4a9d57f9dca6

File: baseline/client/js/jspsych-n-back.js [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+        // initialize state
+        let index = 0;
+        // add event listeners
+        document.addEventListener("keydown", event => {
+            event.preventDefault();
+            if (event.code === "Space") {
+                const data = {
+                    n: trial.n,
+                    sequence: trial.sequence,
+                    response_index: undefined,
```
[the remark this answers is not in the document]
Comment (Floozutter) [reply baseline/client/js/jspsych-n-back.js]: Implemented in 302c0b70c26a86439fa983a2a94a8f1469a2eb85!

File: baseline/client/n-back/n-back.js [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+import "./style.css";
+import introduction_html from "./frag/introduction.html";
+import instruction_html from "./frag/instruction.html";
+import instruction_0_html from "./frag/instruction_0.html";
+import instruction_1_html from "./frag/instruction_1.html";
+import instruction_2_html from "./frag/instruction_2.html";
+import completion_html from "./frag/completion.html";
+
+export class NBack {
+    getTimeline() {
```
[the remark this answers is not in the document]
Comment (noah10) [reply baseline/client/n-back/n-back.js:16]: I see you've addressed the last two bullet points - thanks. Regarding the hardwired digits being replaced with random ones, I see I left something out: That function also needs to accept a number of targets to be embedded into the 15 random digits.

File: baseline/client/js/jspsych-n-back.js [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+                return item !== undefined && item === "1";
+            } else {
+                const a = trial.sequence[index];
+                const b = trial.sequence[index - trial.n];
+                return a !== undefined && b !== undefined && a === b;
+            }
+        };
+        const finish = () => {
+            const now = performance.now();
+            clearInterval(interval);
```
Comment (noah10) [inline baseline/client/js/jspsych-n-back.js:76]: I think you probably need to clear your keyDown listener here as well. Otherwise what happens if the user never presses the space bar during the digit presentation? I tried that during the 0-back trial and found that the keyDown listener fired when I pressed the space bar on the cue screen before the next set of events, meaning that I only ever saw the first element of the sequence in the 1-back trial.

File: baseline/client/js/jspsych-n-back.js [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+            finish_duration: {
+                type: jsPsych.plugins.parameterType.INT,
+                default: 0,
+            },
+        },
+    };
+
+    plugin.trial = (display_element, trial) => {
+        // build and show display HTML
+        {
```
Comment (noah10) [inline baseline/client/js/jspsych-n-back.js:33]: I think I must be missing something, because I'm not clear on what this curly brace is doing here. Can you explain?

File: baseline/client/tests/n-back.test.js [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+        pressKey(" ");  // skip completion
+        const progress = jsPsych.progress();
+        expect(progress.current_trial_global).toBe(progress.total_trials);
+        const relevant = jsPsych.data.get().filter({isRelevant: true}).values();
+        expect(relevant.length).toBeGreaterThan(0);
+    });
+    
+    it("trials are preceded by cues", () => {
+        const timeline = (new NBack()).getTimeline();
+        const trial_indices = Array.from(timeline.entries())
```
Comment (noah10) [inline baseline/client/tests/n-back.test.js:26]: Generally I'm a fan of the functional style, but this took me a while to understand. A couple of alternate suggestions:

More readable:
```
const n_back_trial_indices = [];
for (const [index, trial] of timeline.entries()) {
  if entry.type === 'n-back' {
     nBackIndices.push(index);
  }
}
```

More concise:
```
const reducer = (targetIndices, trial, index) => trial.type === 'n-back' ? targetIndices.concat([index]) : targetIndices;
const n_back_trial_indices = timeline.reduce(reducer, []);
```

**Review by noah10 — CHANGES_REQUESTED** [review]
Looks like progress! There are obviously still some things to be addressed. In addition to the individual comments, I think it will be important to add tests that check (a) if the value of is_correct is as expected (in both the negative and positive cases) and (b) that the display sequence operates as expected if the space bar isn't pushed during a trial and then *is* pushed to progress to the next trial. (That is, that all numbers in the next trial are actually shown, which doesn't happen now.)

File: baseline/client/package.json [hunk 7]
```diff
@@ -5,7 +5,10 @@
     "deploy": "npm run build && npm run test && node deploy.js"
   },
   "dependencies": {
-    "@adp-psych/jspsych": "^6.3.1"
+    "@adp-psych/jspsych": "^6.3.1",
+    "crypto-browserify": "^3.12.0",
```
Comment (noah10) [inline baseline/client/package.json:9]: What are we using this for?

File: baseline/client/package.json [hunk 8]
```diff
@@ -5,7 +5,10 @@
...
+    "random-js": "^2.1.0",
+    "stream-browserify": "^3.0.0"
```
Comment (noah10) [inline baseline/client/package.json:11]: And this one?

**Review by noah10 — DISMISSED** [review]
As per our conversation yesterday, merging this in knowing that the needed changes will come in separate PR's.

**Review by noah10 — APPROVED** [review]
As per our conversation yesterday, merging this in knowing that the needed changes will come in separate PR's.
